File: basics/manacher.cpp
#include <bits/stdc++.h>
using namespace std;
const int inf = 0x3f3f3f3f;
string s;
vector<int> ans;


vector<int> manacher(string s){
  int n = s.size();
  vector<int> d1(n+1);
  vector<int> d2(n+1);
  vector<int> temp(n+1);
  for (int i = 0, l = 0, r = -1, k; i < n; i++){
    k = (i > r ? 1: min(d1[l+r-i], r-i+1));
    while(0 <= i-k && i+k < n && s[i-k] == s[i+k]){
      k++;
    }
    d1[i] = k--;
    if (i+k > r){
      l = i-k;
      r = i+k;
    }
  }
  for (int i = 0, l = 0, r = -1, k; i < n; i++){
    k = (i > r ? 0: min(d2[l+r-i+1], r-i+1));
    while(0 <= i-k-1 && i+k < n && s[i-k-1] == s[i+k]){
      k++;
    }
    d2[i] = k--;
    if (i+k > r){
      l = i-k-1;
      r = i+k;
    }
  }
  for (int i = 0; i < n; i++){
    temp[i] = max((d1[i]*2)-1, d2[i]*2);
  }
  return temp;
}


int main(){
  cin >> s;
  
  ans = manacher(s);

  for (auto ind: ans){
    cout << ind << " ";
  } cout << "\n";

  return 0;
}

File: basics/manacher.py
import sys
s = sys.stdin.readline().strip("\n")


def manacher(s):
  n = len(s)
  d1 = [0 for x in range(n+1)]
  d2 = [0 for x in range(n+1)]
  l = 0
  r = -1
  for i in range(n):
    if i > r:
      k = 1
    else:
      k = min(d1[l+r-i], r-i+1)
    while 0 <= i-k and i+k < n and s[i-k] == s[i+k]:
      k += 1
    d1[i] = k
    k -= 1
    if i+k > r:
      l = i-k
      r = i+k

  l = 0
  r = -1
  for i in range(n):
    if i > r:
      k = 0
    else:
      k = min(d2[l+r-i+1], r-i+1)
    while 0 <= i-k-1 and i+k < n and s[i-k-1] == s[i+k]:
      k += 1
    d2[i] = k
    k -= 1
    if i+k > r:
      l = i-k-1
      r = i+k

  # longest palindrome centered at index i
  # for d2, the center is the right of the 2
  temp = [0 for x in range(n+1)]
  for i in range(0, n):
    temp[i] = max((d1[i]*2)-1, d2[i]*2)
  
  return temp


ans = manacher(s)
print(ans)